test/tstScopedProfileRegion.cpp

```diff
@@ -0,0 +1,96 @@
+/****************************************************************************
+ * Copyright (c) 2017-2021 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <ArborX_DetailsKokkosExtScopedProfileRegion.hpp>
+
+#include <boost/test/unit_test.hpp>
+
+#include <stack>
+#include <string>
+#include <type_traits>
+
+#if (KOKKOS_VERSION >= 30200) // callback registration from within the program
+                              // was added in Kokkkos v3.2
+
+namespace
+{
+std::stack<std::string> arborx_test_region_stack;
+
+// NOTE: cannot use lambdas because they can only be converted to function
+// pointers if they don't capture anything
+void arborx_test_push_region(char const *label)
+{
+  BOOST_TEST_MESSAGE(std::string("push ") + label);
+  arborx_test_region_stack.push(label);
+}
+
+void arborx_test_pop_region()
+{
+  auto const &label = arborx_test_region_stack.top();
+  BOOST_TEST_MESSAGE(std::string("pop ") + label);
+  arborx_test_region_stack.pop();
+}
+
+} // namespace
+
+BOOST_AUTO_TEST_SUITE(KokkosToolsAnnotations)
+
+BOOST_AUTO_TEST_CASE(scoped_profile_region)
+{
+  Kokkos::Tools::Experimental::set_push_region_callback(
+      arborx_test_push_region);
+  Kokkos::Tools::Experimental::set_pop_region_callback(arborx_test_pop_region);
+
+  BOOST_TEST(arborx_test_region_stack.empty());
+
+  // Unnamed guard!  Profile region is popped at the end of the statement.
+  KokkosExt::ScopedProfileRegion("bug");
+
+  BOOST_TEST(arborx_test_region_stack.empty());
+
+  std::string outer_identifier = "outer";
+  KokkosExt::ScopedProfileRegion guard_outer(outer_identifier);
+
+  BOOST_TEST(arborx_test_region_stack.size() == 1);
+  BOOST_TEST(arborx_test_region_stack.top() == outer_identifier);
+
+  {
+    std::string inner_identifier = "inner";
+    KokkosExt::ScopedProfileRegion guard_inner(inner_identifier);
+    BOOST_TEST(arborx_test_region_stack.size() == 2);
+    BOOST_TEST(arborx_test_region_stack.top() == inner_identifier);
+  }
+
+  BOOST_TEST(arborx_test_region_stack.size() == 1);
+  BOOST_TEST(arborx_test_region_stack.top() == outer_identifier);
+
+  arborx_test_region_stack.pop();
```

Comment:
I would not mess with the std::stack manually and just rely on the scope.
```suggestion
  {
    std::string outer_identifier = "outer";
    KokkosExt::ScopedProfileRegion guard_outer(outer_identifier);

    BOOST_TEST(arborx_test_region_stack.size() == 1);
    BOOST_TEST(arborx_test_region_stack.top() == outer_identifier);

    {
      std::string inner_identifier = "inner";
      KokkosExt::ScopedProfileRegion guard_inner(inner_identifier);
      BOOST_TEST(arborx_test_region_stack.size() == 2);
      BOOST_TEST(arborx_test_region_stack.top() == inner_identifier);
    }

    BOOST_TEST(arborx_test_region_stack.size() == 1);
    BOOST_TEST(arborx_test_region_stack.top() == outer_identifier);
  }
```

Reply:
No idea what you mean

Reply:
OK got it.  I would still want to compare the top of the stack.

Reply:
> @dalg24 you want to cleanup the history?

Wasn't planning on it unless someone ask me to.

Reply:
The only difference to use an extra scope instead of `arborx_test_region_stack.pop();`. The check after that can remain the same.
I don't feel strongly about it. It's just a slight preference. If you agree, change it, if not leave it as is.